mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala

```diff
@@ -0,0 +1,690 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.clustering
+
+import scala.collection.{Map, mutable}
+
+import breeze.linalg
+  .{SparseVector => BSV, Vector => BV, any => breezeAny, norm => breezeNorm, sum => breezeSum}
+
+import org.apache.spark.{Logging, SparkException}
+import org.apache.spark.annotation.Since
+import org.apache.spark.mllib.linalg.{Vector, Vectors}
+import org.apache.spark.rdd.RDD
+
+
+/**
+ * This is a divisive hierarchical clustering algorithm based on bisecting k-means algorithm.
+ *
+ * The main idea of this algorithm is based on "A comparison of document clustering techniques",
+ * M. Steinbach, G. Karypis and V. Kumar. Workshop on Text Mining, KDD, 2000.
+ * http://cs.fit.edu/~pkc/classes/ml-internet/papers/steinbach00tr.pdf
+ *
+ * However, we modified it to fit for Spark. This algorithm consists of the two main parts.
+ *
+ * 1. Split clusters until the number of clusters will be enough to build a cluster tree
+ * 2. Build a cluster tree as a binary tree by the splitted clusters
+ *
+ * First, it splits clusters to their children clusters step by step, not considering a cluster
+ * will be included in the final cluster tree or not. That's because it makes the algorithm more
+ * efficient on Spark and splitting a cluster one by one is very slow. It will keep splitting until
+ * the number of clusters will be enough to build a cluster tree. Otherwise, it will stop splitting
+ * when there are no dividable clusters before the number of clusters will be sufficient. And
+ * it calculates the criterions, such as average cost, entropy and so on, for building a cluster
+ * tree in the first part. The criterion means how large the cluster is. That is, the cluster
+ * whose criterion is maximum of all the clusters is the largest cluster.
+ *
+ * Second, it builds a cluster tree as a binary tree by the result of the first part.
+ * First of all, the cluster tree starts with only the root cluster which includes all points.
+ * So, there are two candidates which can be merged to the cluster tree. Those are the children of
+ * the root. Then, it picks up the larger child of the two and merge it to the cluster tree.
+ * After that, there are tree candidates to merge. Those are the smaller child of the root and
+ * the two children of the larger cluster of the root. It picks up the largest cluster of the tree
+ * and merge it to the * cluster tree. Like this, it continues to pick up the largest one of the
+ * candidates and merge it to the cluster tree until the desired number of clusters is reached.
+ *
+ * @param k tne desired number of clusters
+ * @param clusterMap the pairs of cluster and its index as Map
+ * @param maxIterations the number of maximal iterations to split clusters
+ * @param seed a random seed
+ */
+@Since("1.6.0")
+class BisectingKMeans private (
+    private var k: Int,
+    private var clusterMap: Map[BigInt, BisectingClusterNode],
```

Comment:
* remove it from constructor args, move it to `run` method instead
* `BigInt` -> `Long`